ga4gh/client.py

```diff
@@ -29,7 +42,88 @@ def __init__(self, urlPrefix, debugLevel, workarounds, key):
         self._workarounds = workarounds
         self._key = key
 
-    def runRequest(self, request, url, protocolClass, listAttr):
+    # TODO temporary auth solution
+    def _addAuth(self):
+        if self._usingWorkaroundsFor(self.workaroundGoogle):
+            return {'key': self._key}
+        else:
+            return {}
+
+    def _usingWorkaroundsFor(self, workaround):
+        return workaround in self._workarounds
+
+    def _debugResponse(self, jsonString):
+        if self._debugLevel > 1:
+            # TODO use a logging output and integrate with HTTP client more
+            # nicely... also, logging / error handling in general
+            print("json response:")
+            pp = json.dumps(
+                json.loads(jsonString), sort_keys=True, indent=4)
+            print(pp)
+
+    def _checkStatus(self, response):
+        if response.status_code != 200:
+            print(response.text)
+            raise Exception("Url {0} had status_code {1}".format(
+                response.url, response.status_code))
+
+    def _constructUrl(self, url, id=None):
+        fullUrl = posixpath.join(self._urlPrefix, url).format(id=id)
+        if id is None:
+            return fullUrl
+        else:
+            return fullUrl.format(id=id)
+
+    def _updateBytesRead(self, jsonString):
+        self._bytesRead += len(jsonString)
+
+    def _doRequest(self, request, url, requestMethod, id):
+        fullUrl = self._constructUrl(url, id=id)
+        params = self._addAuth()
+        method = getattr(requests, requestMethod)
+        if requestMethod == 'post':
+            headers = {"Content-type": "application/json"}
+            jsonString = request.toJSONString()
+            response = requests.post(
+                fullUrl, jsonString, headers=headers, verify=False,
+                params=params)
+        elif requestMethod == 'get':
+            if request is not None:
+                requestDict = request.toJSONDict()
+                params.update(requestDict)
+            response = requests.get(
+                fullUrl, params=params)
+        self._checkStatus(response)
+        return response
+
+    def _deserializeResponse(self, response, protocolClass):
+        jsonResponseString = response.text
+        self._updateBytesRead(jsonResponseString)
+        self._debugResponse(jsonResponseString)
+        responseObject = protocolClass.fromJSONString(jsonResponseString)
+        return responseObject
+
+    def runRequestSingle(self, rrr):
```

Comment:
Ah, this is a `RunRequestRequest`, so `rrr`. Would `requestContainer` be clearer?